src/datasources/push-notifications-api/firebase-cloud-messaging-api.service.spec.ts

```diff
@@ -0,0 +1,159 @@
+import { fakeJson } from '@/__tests__/faker';
+import { FakeConfigurationService } from '@/config/__tests__/fake.configuration.service';
+import { FakeCacheService } from '@/datasources/cache/__tests__/fake.cache.service';
+import { CacheDir } from '@/datasources/cache/entities/cache-dir.entity';
+import { HttpErrorFactory } from '@/datasources/errors/http-error-factory';
+import { IJwtService } from '@/datasources/jwt/jwt.service.interface';
+import { INetworkService } from '@/datasources/network/network.service.interface';
+import { FirebaseCloudMessagingApiService } from '@/datasources/push-notifications-api/firebase-cloud-messaging-api.service';
+import { faker } from '@faker-js/faker';
+
+const mockNetworkService = jest.mocked({
+  get: jest.fn(),
+  post: jest.fn(),
+} as jest.MockedObjectDeep<INetworkService>);
+
+const mockJwtService = jest.mocked({
+  sign: jest.fn(),
+} as jest.MockedObjectDeep<IJwtService>);
+
+const mockHttpErrorFactory = jest.mocked({
+  from: jest.fn(),
+} as jest.MockedObjectDeep<HttpErrorFactory>);
+
+describe('FirebaseCloudMessagingApiService', () => {
+  let target: FirebaseCloudMessagingApiService;
+  let fakeCacheService: FakeCacheService;
+
+  let pushNotificationsBaseUri: string;
+  let pushNotificationsProject: string;
+  let pushNotificationsServiceAccountClientEmail: string;
+  let pushNotificationsServiceAccountPrivateKey: string;
+
+  beforeEach(() => {
+    jest.resetAllMocks();
+
+    pushNotificationsBaseUri = faker.internet.url({ appendSlash: false });
+    pushNotificationsProject = faker.word.noun();
+    pushNotificationsServiceAccountClientEmail = faker.internet.email();
+    pushNotificationsServiceAccountPrivateKey = faker.string.alphanumeric();
+
+    const fakeConfigurationService = new FakeConfigurationService();
+    fakeConfigurationService.set(
+      'pushNotifications.baseUri',
+      pushNotificationsBaseUri,
+    );
+    fakeConfigurationService.set(
+      'pushNotifications.project',
+      pushNotificationsProject,
+    );
+    fakeConfigurationService.set(
+      'pushNotifications.serviceAccount.clientEmail',
+      pushNotificationsServiceAccountClientEmail,
+    );
+    fakeConfigurationService.set(
+      'pushNotifications.serviceAccount.privateKey',
+      pushNotificationsServiceAccountPrivateKey,
+    );
+
+    fakeCacheService = new FakeCacheService();
+    target = new FirebaseCloudMessagingApiService(
+      mockNetworkService,
+      fakeConfigurationService,
+      fakeCacheService,
+      mockJwtService,
+      mockHttpErrorFactory,
+    );
+  });
+
+  it('it should get an OAuth2 token if not cached, cache it and enqueue a notification', async () => {
+    const oauth2AssertionJwt = faker.string.alphanumeric();
+    const oauth2Token = faker.string.alphanumeric();
+    const oauth2TokenExpiresIn = faker.number.int();
+    const fcmToken = faker.string.alphanumeric();
+    const notification = {
```

Comment:
Nit: I'd add a `firebaseNotificationBuilder` function to generate those payloads.

Reply:
Added in 5c673cf1561f3e7afd444e1b12e88845d629a4ab as well as correcting [the data structure](https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages#Message.FIELDS.data).